katdal/datasources.py

```diff
@@ -623,42 +622,49 @@ def from_url(cls, url, chunk_store='auto', upgrade_flags=True, **kwargs):
             Extra keyword arguments passed to telstate view and chunk store init
         """
-        url_parts = urllib.parse.urlparse(url, scheme='file')
+        u = urllib.parse.urlparse(url, scheme='file')
         # Merge key-value pairs from URL query with keyword arguments
         # of function (the latter takes precedence)
-        url_kwargs = dict(urllib.parse.parse_qsl(url_parts.query))
+        url_kwargs = dict(urllib.parse.parse_qsl(u.query))
         url_kwargs.update(kwargs)
         kwargs = url_kwargs
         # Extract Redis database number if provided
         db = int(kwargs.pop('db', '0'))
-        if url_parts.scheme == 'file':
+        if u.scheme == 'file':
             # RDB dump file
-            telstate = katsdptelstate.TelescopeState()
+            telstate = katsdptelstate.TelescopeState(katsdptelstate.memory.MemoryBackend())
             try:
-                telstate.load_from_file(url_parts.path)
-            except OSError as err:
-                raise DataSourceNotFound(str(err))
-        elif url_parts.scheme == 'redis':
+                telstate.load_from_file(u.path)
+            except OSError as e:
+                raise DataSourceNotFound(str(e))
+        elif u.scheme == 'redis':
             # Redis server
             try:
-                telstate = katsdptelstate.TelescopeState(url_parts.netloc, db)
+                telstate = katsdptelstate.TelescopeState(u.netloc, db)
             except katsdptelstate.ConnectionError as e:
                 raise DataSourceNotFound(str(e))
-        elif url_parts.scheme in {'http', 'https'}:
-            # Open endpoint as an S3 object store (with auth info in kwargs)
-            endpoint = urllib.parse.urlunparse(url_parts[:2] + 4 * ('',))
-            rdb_store = S3ChunkStore.from_url(endpoint, **kwargs)
-            rdb_path = urllib.parse.urlunparse(url_parts[:3] + 3 * ('',))
-            telstate = katsdptelstate.TelescopeState()
-            with rdb_store.request('RDB', 'GET', rdb_path) as response:
-                telstate.load_from_file(io.BytesIO(response.content))
+        elif u.scheme in {'http', 'https'}:
+            # Treat URL prefix as an S3 object store (with auth info in kwargs)
+            store_path = os.path.dirname(os.path.dirname(u.path))
+            store_url = (u.scheme, u.netloc, store_path, '', '', '')
+            store_url = urllib.parse.urlunparse(store_url)
+            # Strip off parameters, query strings and fragments to get basic URL
+            rdb_url = (u.scheme, u.netloc, u.path, '', '', '')
+            rdb_url = urllib.parse.urlunparse(rdb_url)
+            telstate = katsdptelstate.TelescopeState(katsdptelstate.memory.MemoryBackend())
+            try:
+                rdb_store = S3ChunkStore.from_url(store_url, **kwargs)
+                with rdb_store.request('', 'GET', rdb_url) as response:
+                    telstate.load_from_file(io.BytesIO(response.content))
+            except StoreUnavailable as e:
```

Comment:
Why not `ChunkStoreError`? Presumably if the smoke S3ChunkStore construction passes but the rdb file is missing you'll get `ChunkNotFound` rather than `StoreUnavailable`.

Reply:
I've now checked and a missing RDB is indeed a `ChunkStoreError`. I'll catch both.